README.md
# Senior Plataform Apps SDK para Node

Esta biblioteca facilita o uso das APIs de aplicativos da plataforma da Senior Sistemas S/A em projetos Node. Essa mesma biblioteca também está [disponível para Java](https://github.com/dev-senior-com-br/senior-platform-apps-java). A documentação completa das APIs está definida em https://dev.senior.com.br/apis_publicas/.

Atualmente, as APIs disponibilizadas pelo SDK são:
- **Workflow**

Alguns serviços como o HCM e SAM possuem suas próprias bibliotecas, que podem ser encontradas no perfil do [Dev Senior no GitHub](https://github.com/dev-senior-com-br).

## Instalação

Adicione a dependência da biblioteca em seu projeto através do comando:
```bash
npm install --save @seniorsistemas/senior-platform-apps
```

## Ambiente
Atualmente o ambiente padrão para o desenvolvimento é o de homologação (Homologx).
Para fazer a troca do ambiente basta chamar o método `setEnvironment` ou `setUrl` caso queira passar uma url da plataforma diferente das configuradas. Os ambientes principais estão definidos no enum `ENVIRONMENTS` apresentado abaixo:

```javascript
export enum ENVIRONMENTS {
  DEV = 'https://platform-homologx.senior.com.br/t/senior.com.br/bridge/1.0',
  PROD = 'https://api.senior.com.br',
}
```

## Utilização

Para utilizar a biblioteca você precisa apenas se autenticar através da API `authentication` e atribuir para a API o `accessToken` recebido após o processo. Em seguida você poderá utilizar todos os outros serviços disponíveis. O exemplo abaixo demonstra o processo de login e criação de uma solicitação de viagem no BPM através do método `startRequest`:

```javascript
const { PlatformAppsApi } = require('@seniorsistemas/senior-platform-apps');

const username = process.env.PLATFORM_USER;
const password = process.env.PLATFORM_PASS;
// identificador de um processo previamente criado e publicado no Workflow
const processId = process.env.PLATFORM_BPM_PROCESS_ID;

const api = new PlatformAppsApi();

api.authentication.login(username, password).then((resp) => {
  api.accessToken = JSON.parse(resp.body.jsonToken).access_token;
  api.workflow.startRequest({
    processId: processId,
    businessData: {
      destino: 'Curitiba',
      partida: '2021-06-25',
      retorno: '2021-06-28',
      motivo: 'Visita a cliente',
    },
    title: 'Viagem para Curitiba',
  })
  .then(resp => console.log(resp.body))
  .catch(err => console.error(err.response.data));
});
```

### Mudando o ambiente:

```javascript
const { ENVIRONMENTS } = require('@seniorsistemas/senior-core');
api.setEnvironment(ENVIRONMENTS.PROD);
```

## Exemplos

Na pasta [examples](examples) há alguns arquivos com exemplos utilizando a biblioteca. Para executá-los, basta criar as variáveis de ambiente abaixo, sendo que elas podem ser declaradas na máquina ou em um arquivo `.env` na raiz do proejto:

```bash
# Usuário da plataforma
PLATFORM_USER=<seu_usuario_da_plataforma>
# Senha da plataforma
PLATFORM_PASS=<sua_senha_da_plataforma>
```

Alguns exemplos fazem uso de variáveis adicionais, como no caso do Workflow demonstrado acima que utiliza a variável `PLATFORM_BPM_PROCESS_ID` para definir o identificador de um processo previamente publicado na plataforma.

### Exemplos implementados

* [Workflow (BPM)](examples/workflow-examples.js)

## Contribuições

Contribuições ao projeto são bem vindas e podem ser realizadas seguindo nosso [Guia de contribuição](https://dev.senior.com.br/guia-de-contribuicao/).

## Suporte

Criar uma issue [https://github.com/dev-senior-com-br/senior-platform-apps-node/issues](https://github.com/dev-senior-com-br/senior-platform-apps-node/issues)

## License

Copyright © 2020.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
- Adicionado projeto com nome `@seniorsistemas/senior-platform-apps`
- Adicionado resource `workflow.ts` e todos os models necessários para realizar as chamadas ao serviço de Workflow (BPM)
- Adicionado arquivo `workflow-examples.ts` com exemplos de uso do client do Workflow

### Changed
- Alterado README ajustando informações sobre instalação e uso do SDK

## [0.0.1] - 2020-08-16

### Changed

### Fixed

### Removed

### Added

[Unreleased]: https://github.com/dev-senior-com-br/senior-platform-apps-node/compare/v0.0.1...HEAD

[0.0.1]: https://github.com/dev-senior-com-br/senior-platform-apps-node/releases/tag/v0.0.1


examples/workflow-examples.js
require('dotenv').config();
const {
  OrderDirection,
  OrderField,
  PendencyType,
  PlatformAppsApi,
  ProcessFilterType,
  ServiceType,
  SubjectKind,
} = require('../dist');
const { put } = require('axios');

/**
 * Este exemplo utiliza como base um processo previamente modelado e publicado no Workflow
 * de Solicitação de Viagens.
 */

const username = process.env.PLATFORM_USER;
const password = process.env.PLATFORM_PASS;
// identificador de um processo previamente criado e publicado no Workflow
const processId = process.env.PLATFORM_BPM_PROCESS_ID;

const api = new PlatformAppsApi();

api.authentication
  .login({ username, password })
  .then(async resp => {
    console.log('login...\n', resp.body);

    api.accessToken = JSON.parse(resp.body.jsonToken).access_token;

    // inicia uma solicitação, criando uma instância do processo
    // também é possível utilizar o startProcess() para a mesma finalidade
    const processInstanceId = startRequest();

    // realiza todas as operações necessárias para adicionar um anexo em uma solicitação
    addAttachmentToProcessInstance(processInstanceId);

    // recupera lista de todos os processos
    getProcessesList();

    // recupera lista dos processos utilizados recentemente
    getRankingProcesses();

    // recupera os dados do processo de exemplo
    findProcess();

    // recupera a solicitação (instância do processo) criada anteriormente
    getProcessInstance(processInstanceId);

    // recupera a lista de solicitações do processo de exemplo
    getRequestsResume();

    // recupera o histórico da solicitação informada
    getRequestHistoryTimeline(processInstanceId);

    // lista as solicitações de aplicações terceiras
    getThirdPartyRequestByStatus();

    // recupera todos os possíveis responsáveis
    getSubjects();

    // recupera todos os possíveis responsáveis para a próxima tarefa
    getNextSubject(processInstanceId);
    getNextSubjectFromInitialTask();

    // recupera lista de tarefas filtradas pelo 'processId' e ordenadas pela data de expiração
    searchTasks();

    // recupera todas as pendências do usuário autenticado
    getMyPendencies();

    // recupera as ações possíveis para uma pendência
    getPendencyProcessActions(processInstanceId);

    // altera o usuário responsável pela solicitação
    changePendencyUser(processInstanceId);

    // responde uma pendência ou uma lista de pendências
    responsePendency(processInstanceId);
    batchPendenciesResponse(processInstanceId);

    // cancela a solicitação iniciada anteriormente
    cancelProcessInstance(processInstanceId);
  })
  .catch(err => {
    console.error(err);
  });

function startRequest() {
  api.workflow
    .startRequest({
      processId: processId,
      businessData: {
        destino: 'Curitiba',
        partida: '2021-06-25',
        retorno: '2021-06-28',
        motivo: 'Visita a cliente',
      },
      title: 'Viagem para Curitiba',
      actionToExecute: 'Analisar',
    })
    .then(resp => console.log('startRequest...\n', resp.body))
    .catch(err => console.error('startRequest...\n', err.response.data));
}

function startProcess() {
  api.workflow
    .startProcess({
      businessData:
        '{"root": {"destino":"Curitiba","partida":"2021-06-25","retorno":"2021-06-28","motivo":"Visita a cliente"}}',
      flowExecutionData: { actionToExecute: 'Cotação' },
      processId: processId,
      processVersion: 1,
      requester: 'senior',
      title: 'Viagem para Curitiba',
    })
    .then(resp => console.log('startProcess...\n', resp.body))
    .catch(err => console.error('startProcess...\n', err.response.data));
}

function getProcessesList() {
  api.workflow
    .getProcessesList({
      serviceAction: ServiceType.AllProcesses,
      onlyActiveProcesses: false,
      processFilterType: ProcessFilterType.All,
    })
    .then(resp => console.log('getProcessesList...\n', resp.body))
    .catch(err => console.error('getProcessesList...\n', err.response.data));
}

function getRankingProcesses() {
  api.workflow
    .getRankingProcesses({
      start: '2020-01-01T00:00:00.00Z',
      limit: 5,
    })
    .then(resp => console.log('getRankingProcesses...\n', resp.body))
    .catch(err => console.error('getRankingProcesses...\n', err.response.data));
}

function findProcess() {
  api.workflow
    .findProcess({
      id: processId,
    })
    .then(resp => console.log('findProcess...\n', resp.body))
    .catch(err => console.error('findProcess...\n', err.response.data));
}

function getProcessInstance(processInstanceId) {
  api.workflow
    .getProcessInstance({
      processInstance: processInstanceId,
    })
    .then(resp => console.log('getProcessInstance...\n', resp.body))
    .catch(err => console.error('getProcessInstance...\n', err.response.data));
}

function getRequestsResume() {
  api.workflow
    .getRequestsResume({
      filterProcess: [processId],
      orders: [{ key: 'processInstanceId', value: OrderDirection.DESC }],
    })
    .then(resp => console.log('getRequestsResume...\n', resp.body))
    .catch(err => console.error('getRequestsResume...\n', err.response.data));
}

function getRequestHistoryTimeline(processInstanceId) {
  api.workflow
    .getRequestHistoryTimeline({
      processInstanceID: processInstanceId,
    })
    .then(resp => console.log('getRequestHistoryTimeline...\n', resp.body))
    .catch(err => console.error('getRequestHistoryTimeline...\n', err.response.data));
}

function getThirdPartyRequestByStatus() {
  api.workflow
    .getThirdPartyRequestByStatus({
      processKey: 'process_key',
      status: PendencyType.Pending,
    })
    .then(resp => console.log('getThirdPartyRequestByStatus...\n', resp.body))
    .catch(err => console.error('getThirdPartyRequestByStatus...\n', err.response.data));
}

function getSubjects() {
  api.workflow
    .getSubjects({
      type: SubjectKind.User,
    })
    .then(resp => console.log('getSubjects...\n', resp.body))
    .catch(err => console.error('getSubjects...\n', err.response.data));
}

function getNextSubject(processInstanceId) {
  api.workflow
    .getNextSubject({
      serviceFlowToken: {
        activityId: 1,
        processInstanceID: processInstanceId,
        step: 1,
      },
      sequenceName: 'Prosseguir',
    })
    .then(resp => console.log('getNextSubject...\n', resp.body))
    .catch(err => console.error('getNextSubject...\n', err.response.data));
}

function getNextSubjectFromInitialTask() {
  api.workflow
    .getNextSubjectFromInitialTask({
      processId: processId,
      processVersion: 10,
      sequenceName: 'Prosseguir',
    })
    .then(resp => console.log('getNextSubjectFromInitialTask...\n', resp.body))
    .catch(err => console.error('getNextSubjectFromInitialTask...\n', err.response.data));
}

function searchTasks() {
  api.workflow
    .searchTasks({
      orders: [{ field: OrderField.EXPIRATION_DATE, direction: OrderDirection.ASC }],
      filter: { processes: [processId] },
    })
    .then(resp => console.log('searchTasks...\n', resp.body))
    .catch(err => console.error('searchTasks...\n', err.response.data));
}

function getMyPendencies() {
  api.workflow
    .getMyPendencies({
      pendencyRequestParameters: {
        type: PendencyType.Pending,
        processFilterType: ProcessFilterType.All,
      },
    })
    .then(resp => console.log('getMyPendencies...\n', resp.body))
    .catch(err => console.error('getMyPendencies...\n', err.response.data));
}

function getPendencyProcessActions(processInstanceId) {
  api.workflow
    .getPendencyProcessActions({
      serviceFlowToken: {
        activityId: 10,
        processInstanceID: processInstanceId,
        step: 30,
      },
    })
    .then(resp => console.log('getPendencyProcessActions...\n', resp.body))
    .catch(err => console.error('getPendencyProcessActions...\n', err.response.data));
}

function changePendencyUser(processInstanceId) {
  api.workflow
    .changePendencyUser({
      serviceFlowTokens: {
        activityId: 10,
        processInstanceID: processInstanceId,
        step: 30,
      },
      subject: {
        subjectKind: SubjectKind.User,
        userCode: 1,
        name: 'Exemplo',
      },
    })
    .then(resp => console.log('changePendencyUser...\n', resp.body))
    .catch(err => console.error('changePendencyUser...\n', err.response.data));
}

/**
 * Realiza as operações newAttachment, commitAttachment e linkAttachments, que
 * resulta na criação e relação de um anexo com uma solicitação.
 */
function addAttachmentToProcessInstance(processInstanceId) {
  api.workflow
    .newAttachment({ name: 'file2.txt' })
    .then(async resp => {
      console.log('newAttachment...', resp.body);
      const attachmentId = resp.body.attachment.id;

      await put(resp.body.uploadUrl, 'My text file', { headers: [{ 'Content-type': 'text/plain' }] });

      api.workflow
        .commitAttachment({ id: attachmentId })
        .then(resp => {
          console.log('commitAttachment...\n', resp.body);

          api.workflow
            .linkAttachments({
              ids: [attachmentId],
              processInstance: processInstanceId,
            })
            .then(resp => console.log('linkAttachments...\n', resp.body))
            .catch(err => console.error('linkAttachments...\n', err.response.data));
        })
        .catch(err => {
          console.error('commitAttachment...\n', err.response.data);
        });
    })
    .catch(err => {
      console.error('newAttachment...\n', err.response.data);
    });
}

function responsePendency(processInstanceId) {
  api.workflow
    .responsePendency({
      responseData: {
        businessData: '{ "aprovado:" true }',
        flowExecutionData: {
          actionToExecute: 'Cotação',
        },
      },
      serviceFlowToken: {
        activityId: 10,
        processInstanceID: processInstanceId,
        step: 30,
      },
    })
    .then(resp => console.log('responsePendency...\n', resp.body))
    .catch(err => console.error('responsePendency...\n', err.response.data));
}

function batchPendenciesResponse(processInstanceId) {
  api.workflow
    .batchPendenciesResponse({
      flowTokens: [
        {
          processInstanceId: processInstanceId,
          step: 30,
          activityId: 10,
        },
      ],
      authorization: api.accessToken,
    })
    .then(resp => console.log('batchPendenciesResponse...\n', resp.body))
    .catch(err => console.error('batchPendenciesResponse...\n', err.response.data));
}

function cancelProcessInstance(processInstanceId) {
  api.workflow
    .cancelProcessInstance({
      ids: [processInstanceId],
      reason: 'Voo cancelado',
    })
    .then(resp => console.log('cancelProcessInstance...\n', resp.body))
    .catch(err => console.error('cancelProcessInstance...\n', err.response.data));
}
